File: daq/slc/runcontrol/src/RCState.cc
/**************************************************************************
 * basf2 (Belle II Analysis Software Framework)                           *
 * Author: The Belle II Collaboration                                     *
 *                                                                        *
 * See git log for contributors and copyright holders.                    *
 * This file is licensed under LGPL-3.0, see LICENSE.md.                  *
 **************************************************************************/
#include "daq/slc/runcontrol/RCState.h"

using namespace Belle2;

const RCState RCState::OFF_S(1, "OFF");
const RCState RCState::NOTREADY_S(2, "NOTREADY");
const RCState RCState::READY_S(3, "READY");
const RCState RCState::RUNNING_S(4, "RUNNING");
const RCState RCState::PAUSED_S(5, "PAUSED");
const RCState RCState::LOADING_TS(6, "LOADING");
const RCState RCState::STARTING_TS(7, "STARTING");
const RCState RCState::STOPPING_TS(8, "STOPPING");
const RCState RCState::CONFIGURING_TS(9, "CONFIGURING");
const RCState RCState::ERROR_ES(10, "ERROR");
const RCState RCState::FATAL_ES(11, "FATAL");
const RCState RCState::BOOTING_RS(12, "BOOTING");
const RCState RCState::RECOVERING_RS(13, "RECOVERING");
const RCState RCState::ABORTING_RS(14, "ABORTING");

const RCState& RCState::operator=(const std::string& label)
{
  if (label == NOTREADY_S.getLabel()) *this = NOTREADY_S;
  else if (label == READY_S.getLabel()) *this = READY_S;
  else if (label == RUNNING_S.getLabel()) *this = RUNNING_S;
  else if (label == PAUSED_S.getLabel()) *this = PAUSED_S;
  else if (label == LOADING_TS.getLabel()) *this = LOADING_TS;
  else if (label == STARTING_TS.getLabel()) *this = STARTING_TS;
  else if (label == STOPPING_TS.getLabel()) *this = STOPPING_TS;
  else if (label == CONFIGURING_TS.getLabel()) *this = CONFIGURING_TS;
  else if (label == ERROR_ES.getLabel()) *this = ERROR_ES;
  else if (label == FATAL_ES.getLabel()) *this = FATAL_ES;
  else if (label == RECOVERING_RS.getLabel()) *this = RECOVERING_RS;
  else if (label == ABORTING_RS.getLabel()) *this = ABORTING_RS;
  else if (label == BOOTING_RS.getLabel()) *this = BOOTING_RS;
  else *this = Enum::UNKNOWN;
  return *this;
}

const RCState& RCState::operator=(const char* label)
{
  if (label != NULL)  *this = std::string(label);
  else *this = Enum::UNKNOWN;
  return *this;
}

const RCState& RCState::operator=(int id)
{
  if (id == NOTREADY_S.getId()) *this = NOTREADY_S;
  else if (id == READY_S.getId()) *this = READY_S;
  else if (id == RUNNING_S.getId()) *this = RUNNING_S;
  else if (id == PAUSED_S.getId()) *this = PAUSED_S;
  else if (id == LOADING_TS.getId()) *this = LOADING_TS;
  else if (id == STARTING_TS.getId()) *this = STARTING_TS;
  else if (id == STOPPING_TS.getId()) *this = STOPPING_TS;
  else if (id == CONFIGURING_TS.getId()) *this = CONFIGURING_TS;
  else if (id == ERROR_ES.getId()) *this = ERROR_ES;
  else if (id == FATAL_ES.getId()) *this = FATAL_ES;
  else if (id == RECOVERING_RS.getId()) *this = RECOVERING_RS;
  else if (id == ABORTING_RS.getId()) *this = ABORTING_RS;
  else if (id == BOOTING_RS.getId()) *this = BOOTING_RS;
  else *this = Enum::UNKNOWN;
  return *this;
}

RCState RCState::next() const
{
  if (*this == LOADING_TS) return READY_S;
  else if (*this == STARTING_TS) return RUNNING_S;
  else if (*this == STOPPING_TS) return READY_S;
  else if (*this == RECOVERING_RS) return READY_S;
  else if (*this == ABORTING_RS) return NOTREADY_S;
  else if (*this == BOOTING_RS) return NOTREADY_S;
  else if (*this == CONFIGURING_TS) return NOTREADY_S;
  else return Enum::UNKNOWN;
}

File: daq/slc/extra/javalib/css/org.csstudio.belle2.daq/src/org/belle2/daq/runcontrol/RCState.java
package org.belle2.daq.runcontrol;

import org.belle2.daq.nsm.NSMState;

public class RCState extends NSMState {

    final static public RCState OFF_S = new RCState(1, "OFF");
    final static public RCState NOTREADY_S = new RCState(2, "NOTREADY");
    final static public RCState READY_S = new RCState(3, "READY");
    final static public RCState RUNNING_S = new RCState(4, "RUNNING");
    final static public RCState PAUSED_S = new RCState(5, "PAUSED");
    final static public RCState LOADING_TS = new RCState(6, "LOADING");
    final static public RCState STARTING_TS = new RCState(7, "STARTING");
    final static public RCState STOPPING_TS = new RCState(8, "STOPPING");
    final static public RCState BOOTING_TS = new RCState(9, "BOOTING");
    final static public RCState ERROR_ES = new RCState(10, "ERROR");
    final static public RCState FATAL_ES = new RCState(11, "FATAL");
    final static public RCState RECOVERING_RS = new RCState(12, "RECOVERING");
    final static public RCState ABORTING_RS = new RCState(13, "ABORTING");

    public static RCState get(int id) {
        RCState state = new RCState();
        state.copy(id);
        return state;
    }

    public boolean isStable() {
        return getId() > 0 && getId() <= PAUSED_S.getId();
    }

    public boolean isTransition() {
        return getId() > PAUSED_S.getId() && getId() <= STOPPING_TS.getId();
    }

    public boolean isError() {
        return getId() >= ERROR_ES.getId() && getId() <= FATAL_ES.getId();
    }

    public boolean isRecover() {
        return getId() >= RECOVERING_RS.getId() && getId() <= ABORTING_RS.getId();
    }

    public RCState() {
        this(UNKNOWN.getId(), UNKNOWN.getLabel());
    }

    protected RCState(int id, String label) {
        super(id, label);
    }

    public RCState(RCState state) {
        super(state.getId(), state.getLabel());
    }

    public RCState(String state) {
        copy(state);
    }

    @Override
    public void copy(String label) {
        if (label.equals(OFF_S.getLabel())) {
            copy(OFF_S);
        } else if (label.equals(NOTREADY_S.getLabel())) {
            copy(NOTREADY_S);
        } else if (label.equals(READY_S.getLabel())) {
            copy(READY_S);
        } else if (label.equals(RUNNING_S.getLabel())) {
            copy(RUNNING_S);
        } else if (label.equals(PAUSED_S.getLabel())) {
            copy(PAUSED_S);
        } else if (label.equals(LOADING_TS.getLabel())) {
            copy(LOADING_TS);
        } else if (label.equals(STARTING_TS.getLabel())) {
            copy(STARTING_TS);
        } else if (label.equals(STOPPING_TS.getLabel())) {
            copy(STOPPING_TS);
        } else if (label.equals(BOOTING_TS.getLabel())) {
            copy(BOOTING_TS);
        } else if (label.equals(ERROR_ES.getLabel())) {
            copy(ERROR_ES);
        } else if (label.equals(FATAL_ES.getLabel())) {
            copy(FATAL_ES);
        } else if (label.equals(RECOVERING_RS.getLabel())) {
            copy(RECOVERING_RS);
        } else if (label.equals(ABORTING_RS.getLabel())) {
            copy(ABORTING_RS);
        } else {
            copy(UNKNOWN);
        }
    }

    public RCState next() {
        if (equals(LOADING_TS)) {
            return READY_S;
        } else if (equals(STARTING_TS)) {
            return RUNNING_S;
        } else if (equals(STOPPING_TS)) {
            return READY_S;
        } else {
            return new RCState();
        }
    }

    @Override
    public void copy(int id) {
        if (id == OFF_S._id) {
            copy(OFF_S);
        } else if (id == NOTREADY_S._id) {
            copy(NOTREADY_S);
        } else if (id == READY_S._id) {
            copy(READY_S);
        } else if (id == RUNNING_S._id) {
            copy(RUNNING_S);
        } else if (id == PAUSED_S._id) {
            copy(PAUSED_S);
        } else if (id == LOADING_TS._id) {
            copy(LOADING_TS);
        } else if (id == STARTING_TS._id) {
            copy(STARTING_TS);
        } else if (id == STOPPING_TS._id) {
            copy(STOPPING_TS);
        } else if (id == BOOTING_TS._id) {
            copy(BOOTING_TS);
        } else if (id == ERROR_ES._id) {
            copy(ERROR_ES);
        } else if (id == FATAL_ES._id) {
            copy(FATAL_ES);
        } else if (id == RECOVERING_RS._id) {
            copy(RECOVERING_RS);
        } else if (id == ABORTING_RS._id) {
            copy(ABORTING_RS);
        } else {
            copy(UNKNOWN);
        }
    }

}
